ssh-pkcs11.c

```diff
@@ -819,6 +941,138 @@ pkcs11_fetch_ecdsa_pubkey(struct pkcs11_provider *p, CK_ULONG slotidx,
+{
+#ifdef NID_ED25519
+	const unsigned char *p = param;
+	ASN1_OBJECT *pObj = d2i_ASN1_OBJECT(NULL, &p, len);
+
+	/* Edwards EC public keys only support the use of the curveName
+	 * selection to specify a curve name as defined in [RFC 8032]
+	 * and the use of the oID selection to specify a curve through
+	 * an EdDSA algorithm as defined in [RFC 8410]. */
+	if (pObj != NULL) {
+		int nid = OBJ_obj2nid(pObj);
+		ASN1_OBJECT_free(pObj);
+		return (nid == NID_ED25519) ? (0) : (-1);
+	} else {
+		ASN1_PRINTABLESTRING *curve_name = NULL;
+
+		p = param;
+		curve_name = d2i_ASN1_PRINTABLESTRING(NULL, &p, len);
+		if (curve_name && (strcmp((char *)curve_name->data, "edwards25519") == 0)) {
+			ASN1_PRINTABLESTRING_free(curve_name);
+			return (0);
+		}
+
+		if (curve_name) {
+			ASN1_PRINTABLESTRING_free(curve_name);
+			return (-1);
+		}
+		/* Unsupported (yet) parameters format.
+		 * If the key is invalid, we get the failure on signature verification. */
+		logit("unsupported edwards parameters format");
+		return (0);
+	}
+#else
+	/* Cannot really validate.
+	 * If the key is invalid, we get the failure on signature verification. */
+	logit("Cannot validate edwards parameters");
+	return (0);
+#endif
+}
+
+
+static struct sshkey *
+pkcs11_fetch_eddsa_pubkey(struct pkcs11_provider *p, CK_ULONG slotidx,
+    CK_OBJECT_HANDLE *obj)
+{
+	CK_ATTRIBUTE		 key_attr[3];
+	CK_SESSION_HANDLE	 session;
+	CK_FUNCTION_LIST	*f = NULL;
+	CK_RV			 rv;
+	struct sshkey		*key = NULL;
+	int			 i;
+	u_char			*ed25519_pk = NULL;
+
+	memset(&key_attr, 0, sizeof(key_attr));
+	key_attr[0].type = CKA_ID;
+	key_attr[1].type = CKA_EC_POINT;
+	key_attr[2].type = CKA_EC_PARAMS;
+
+	session = p->slotinfo[slotidx].session;
+	f = p->function_list;
+
+	/* figure out size of the attributes */
+	rv = f->C_GetAttributeValue(session, *obj, key_attr, 3);
+	if (rv != CKR_OK) {
+		error("C_GetAttributeValue failed: %lu", rv);
+		return (NULL);
+	}
+
+	/*
+	 * Allow CKA_ID (always first attribute) to be empty, but
+	 * ensure that none of the others are zero length.
+	 * XXX assumes CKA_ID is always first.
+	 */
+	if (key_attr[1].ulValueLen == 0 ||
+	    key_attr[2].ulValueLen == 0) {
+		error("invalid attribute length");
+		return (NULL);
+	}
+
+	/* allocate buffers for attributes */
+	for (i = 0; i < 3; i++)
+		if (key_attr[i].ulValueLen > 0)
+			key_attr[i].pValue = xcalloc(1, key_attr[i].ulValueLen);
+
+	/* retrieve ID, public point and curve parameters of Edwards key */
+	rv = f->C_GetAttributeValue(session, *obj, key_attr, 3);
+	if (rv != CKR_OK) {
+		error("C_GetAttributeValue failed: %lu", rv);
+		goto fail;
+	}
+
+	if (key_attr[1].ulValueLen != crypto_sign_ed25519_PUBLICKEYBYTES) {
+		error("CKA_EC_POINT invalid length");
+		goto fail;
+	}
+
+	ed25519_pk = malloc(key_attr[1].ulValueLen);
+	if (ed25519_pk == NULL) {
+		error("malloc failed");
+		goto fail;
+	}
+	memcpy(ed25519_pk, key_attr[1].pValue, key_attr[1].ulValueLen);
```

Comment:
shouldn't `CKA_EC_POINT` be a asn.1-encoded octet string? that's the behaviour currently [implemented](https://github.com/OpenSC/OpenSC/blob/35a8a1d7e14aa666c6fc14643ade32dcea280379/src/pkcs11/framework-pkcs15.c#L5634) in OpenSC, and seems to match [section 2.3.5 of pkcs#11 3.0](https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/csprd01/pkcs11-curr-v3.0-csprd01.html#_Toc10560871).

Reply:
Could you please test it against OpenSC?

Reply:
sure! here's the relevant debug info:

```
P:134583; T:0x139995435037632 13:57:56.460 [opensc-pkcs11] pkcs15-pubkey.c:809:sc_pkcs15_encode_pubkey_eddsa: called
P:134583; T:0x139995435037632 13:57:56.460 [opensc-pkcs11] asn1.c:1796:asn1_encode_entry: encoding 'pubkey'
P:134583; T:0x139995435037632 13:57:56.460 [opensc-pkcs11] asn1.c:1802:asn1_encode_entry: type=4, tag=0x04, parm=0x603000025690, len=32
P:134583; T:0x139995435037632 13:57:56.460 [opensc-pkcs11] asn1.c:1994:asn1_encode_entry: length of encoded item=34
P:134583; T:0x139995435037632 13:57:56.460 [opensc-pkcs11] pkcs15-pubkey.c:814:sc_pkcs15_encode_pubkey_eddsa: returning with: 0 (Success)
P:134583; T:0x139995435037632 13:57:56.460 [opensc-pkcs11] pkcs11-object.c:263:C_GetAttributeValue: Object 106102872082208: CKA_EC_POINT = 0420E16AC1E39E744FA1160ADCC1AA8EC27E6592E61B13664B8E959C4B379FC9
```

(note the 0420 sequence)

Reply:
Thanks! Will test and add support of this format too.

Reply:
@martelletto Sorry, I don't get. The string `0420E16AC1E39E744FA1160ADCC1AA8EC27E6592E61B13664B8E959C4B379FC9` seems to be too short: 04 indicates OCTET STRING, 20 indicates 32 bytes length, but the whole dump is 64 bytes, so it looks like it is cut

Reply:
@beldmit That's indeed confusing. The output was cropped by OpenSC's debug routines; see [`sc_pkcs11_print_value()`](https://github.com/OpenSC/OpenSC/blob/master/src/pkcs11/debug.c#L213-L234) and [`DUMP_TEMPLATE_MAX`](https://github.com/OpenSC/OpenSC/blob/master/src/pkcs11/debug.c#L28). Cranking `DUMP_TEMPLATE_MAX` to 128 yields the complete blob:

```
P:49231; T:0x140256057680832 15:55:21.387 [opensc-pkcs11] pkcs11-object.c:263:C_GetAttributeValue: Object 106102872082208: CKA_EC_POINT = 0420E16AC1E39E744FA1160ADCC1AA8EC27E6592E61B13664B8E959C4B379FC96EBF
```

Reply:
Thanks! Fixup pushed.